packages/core/src/routes/admin-user/mfa-verifications.ts

```diff
@@ -0,0 +1,159 @@
+import { MfaFactor, userMfaVerificationResponseGuard } from '@logto/schemas';
+import { generateStandardId, getUserDisplayName } from '@logto/shared';
+import { authenticator } from 'otplib';
+import qrcode from 'qrcode';
+import { object, string, z } from 'zod';
+
+import RequestError from '#src/errors/RequestError/index.js';
+import koaGuard from '#src/middleware/koa-guard.js';
+import assertThat from '#src/utils/assert-that.js';
+import { transpileUserMfaVerifications } from '#src/utils/user.js';
+
+import { generateBackupCodes } from '../interaction/utils/backup-code-validation.js';
+import { generateTotpSecret } from '../interaction/utils/totp-validation.js';
+import type { AuthedRouter, RouterInitArgs } from '../types.js';
+
+export default function adminUserMfaVerificationsRoutes<T extends AuthedRouter>(
+  ...args: RouterInitArgs<T>
+) {
+  const [router, { queries, libraries }] = args;
+  const {
+    users: { findUserById, updateUserById },
+  } = queries;
+
+  router.get(
+    '/users/:userId/mfa-verifications',
+    koaGuard({
+      params: object({ userId: string() }),
+      response: userMfaVerificationResponseGuard,
+      status: [200, 404],
+    }),
+    async (ctx, next) => {
+      const user = await findUserById(ctx.guard.params.userId);
+      ctx.body = transpileUserMfaVerifications(user.mfaVerifications);
+      return next();
+    }
+  );
+
+  router.post(
+    '/users/:userId/mfa-verifications',
+    koaGuard({
+      params: object({ userId: string() }),
+      body: z.object({
+        type: z.literal(MfaFactor.TOTP).or(z.literal(MfaFactor.BackupCode)),
+      }),
+      response: z.discriminatedUnion('type', [
+        z.object({
+          type: z.literal(MfaFactor.TOTP),
+          secret: z.string(),
+          secretQrCode: z.string(),
+        }),
+        z.object({
+          type: z.literal(MfaFactor.BackupCode),
+          codes: z.string().array(),
+        }),
+      ]),
+      status: [200, 404],
+    }),
+    async (ctx, next) => {
+      const { id, mfaVerifications, username, primaryEmail, primaryPhone, name } =
+        await findUserById(ctx.guard.params.userId);
+
+      const { type } = ctx.guard.body;
+
+      if (type === MfaFactor.TOTP) {
+        // A user can only bind one TOTP factor
+        assertThat(
+          mfaVerifications.every(({ type }) => type !== MfaFactor.TOTP),
+          new RequestError({
+            code: 'user.totp_already_in_use',
+            status: 422,
+          })
+        );
+
+        const secret = generateTotpSecret();
+        const service = ctx.URL.hostname;
+        const user = getUserDisplayName({ username, primaryEmail, primaryPhone, name });
+        const keyUri = authenticator.keyuri(user, service, secret);
+        await updateUserById(id, {
+          mfaVerifications: [
+            ...mfaVerifications,
+            { type, key: secret, id: generateStandardId(), createdAt: new Date().toISOString() },
+          ],
+        });
```

Comment:
Feels like this kind of logic can be extracted as a common library method.

Reply:
Extracted as `userLibrary.addUserMfaVerification`. But this is not applied to "submit interaction" because it also update some other user infos.